src/transfer.jl

```diff
@@ -1,120 +1,210 @@
         end
-        if b > r0
-            #if the ray never leaves the model atmosphere, include 
-            #the contribution from the other side of the star
-            τ_prime = τ[end:end, :] .+ [cumsum(reverse(dτ[2:end, :], dims=1), dims=1) 
-                                        ; zeros(size(τ[end:end, :]))]
-            for j in 1:size(τ, 2) #Iterate over wavelength
-                I[j, μ_ind] += transfer_integral(τ_prime[:, j], S[1:i, j]; plane_parallel=false)
-            end
-        else #otherwise assume I=S at atmosphere lower boundary
-            I[:, μ_ind] += @. exp(-τ[end, :]) * S[end, :]
+        cumulative_trapezoid_rule!(τ_λ, log_τ_ref, integrand, i)
+        I[λ_ind, μ_ind] = ray_transfer_integral(view(τ_λ,1:i), view(S,1:i,λ_ind))
+
+        if i < length(radii)
+            #if the ray never leaves the model atmosphere, include the contribution from the 
+            #other side of the star.  Calculate the optical depths you get by reversing the τ_λ.
+            #This is less accurate than actually integrating to find τ, but the effect is small.
+            #This should probably by audited for off-by-one errors.  It's also inneficient.
+            τ_prime = τ_λ[i] .+ [cumsum(reverse(diff(view(τ_λ,1:i)))) ; 0]
+            I[λ_ind, μ_ind] += ray_transfer_integral(view(τ_prime, 1:i), view(S,1:i,λ_ind))
+        else #otherwise assume I=S at atmosphere lower boundary.  This is a _tiny_ effect.
+            I[λ_ind, μ_ind] += exp(-τ_λ[end]) * S[end, λ_ind]
         end
     end
+    #calculate 2π∫μIdμ to get astrophysical flux
+    #this is likely not the most efficient way to do this.
+    F = 2π * [Korg.trapezoid_rule(μ_surface_grid, μ_surface_grid .* I) for I in eachrow(I)]
+    I, F
+end
+
+
+"""
+    all_mu_transfer_integral(τ, S)
+
+Compute exactly the solution to the transfer integral obtained be linearly interpolating the source 
+function, `S` across optical depths `τ`, without approximating the factor of E₂(τ).
+"""
+function all_mu_transfer_integral(τ, S)
+    I = 0
+    for i in 1:length(τ)-1
+        @inbounds m = (S[i+1] - S[i])/(τ[i+1] - τ[i])
+        @inbounds b = S[i] - m*τ[i]
+        @inbounds I += (_plane_parallel_approximate_transfer_integral(τ[i+1], m, b) - 
+                        _plane_parallel_approximate_transfer_integral(τ[i], m, b))
+    end
     I
 end
 
 """
-    trapezoid_rule(xs, fs)
+    ray_transfer_integral(τ, S)
+
+Compute exactly the solution to the transfer integral obtained be linearly interpolating the source 
+function, `S` across optical depths `τ`, without approximating the factor of exp(-τ)
+"""
+function ray_transfer_integral(τ, S)
+    I = 0
+    for i in 1:length(τ)-1
+        @inbounds m = (S[i+1] - S[i])/(τ[i+1] - τ[i])
+        @inbounds b = S[i] - m*τ[i]
+        @inbounds I += (_ray_approximate_transfer_integral(τ[i+1], m, b) - 
+                        _ray_approximate_transfer_integral(τ[i], m, b))
```

Comment:
Since `ray_transfer_integral` is the most expensive part of spherical transfer, it's probably worth exploring whether `_ray_approximate_transfer_integral` is being correctly inlined at some point. 

I'm a little worried that the compiler might not realize that it can rewrite this for loop as follows (and evaluating twice as many exponential functions as it needs to):

```julia
    next_exp_negτ = exp(-τ[1])
    for i in 1:length(τ)-1
        @inbounds m = (S[i+1] - S[i])/(τ[i+1] - τ[i])
        @inbounds b = S[i] - m*τ[i]
        cur_exp_negτ = next_exp_negτ
        @inbounds next_exp_negτ = exp(-τ[i+1])
        @inbounds I += (-next_exp_negτ * (m*τ[i+1] + b + m) +  
                        cur_exp_negτ * (m*τ[i] + b + m))
    end
```

If you have a spare moment, would you mind copying the preceding snippet into the code and seeing if the code gets any faster? If it does speed up the code, we might be able to reproduce the same effect with `@inline` (although I remember that the inline macro was temperamental, so it would be better to test things by directly overwriting the loop)

Reply:
Actually, if we wanted to take this to the logical extreme, one could imagine entirely replacing `ray_transfer_integral` with the following (but I'm not sure it's worth doing given that it's not obvious where this comes from).

```julia
function ray_transfer_integral(τ, S)
    I = 0
    next_exp_negτ = exp(-τ[1])
    for i in 1:length(τ)-1
        @inbounds m = (S[i+1] - S[i])/(τ[i+1] - τ[i])
        cur_exp_negτ = next_exp_negτ
        @inbounds next_exp_negτ = exp(-τ[i+1])
        @inbounds I += (-next_exp_negτ * (S[i+1] + m) +  
                        cur_exp_negτ * (S[i] + m))
    end
    I
end
```

Reply:
I think this is a good idea.  It also concords nicely with my (arguably wrong) preference for few, more complicated, functions.

Reply:
Wait, don't you also need the contribution from `b` at the first and last indices?

Reply:
Unless I did something incorrectly, I don't think so. I literally just substituted the equation for b in and things cancelled out, a lot

Reply:
Ah, I see. Fixed.

Reply:
This is about a 20% speedup, btw.  (Sorry for the message spam.)

Reply:
Okay so here's the sequence of algebraic steps I followed. I started with this function:

```julia
function ray_transfer_integral(τ, S)
    I = 0
    next_exp_negτ = exp(-τ[1])
    for i in 1:length(τ)-1
        @inbounds m = (S[i+1] - S[i])/(τ[i+1] - τ[i])
        @inbounds b = S[i] - m*τ[i]
        cur_exp_negτ = next_exp_negτ
        @inbounds next_exp_negτ = exp(-τ[i+1])
        @inbounds I += (-next_exp_negτ * (m*τ[i+1] + b + m) +  
                        cur_exp_negτ * (m*τ[i] + b + m))
    end
    I
```

1. I literally just extracted the following line:
```julia
    I += (-next_exp_negτ * (m*τ[i+1] + b + m) +  
          cur_exp_negτ * (m*τ[i] + b + m))
```
2. I substituted in `b = S[i] - m*τ[i]`
```julia
    I += (-next_exp_negτ * (m*τ[i+1] + (S[i] - m*τ[i]) + m) +  
          cur_exp_negτ * (m*τ[i] + (S[i] - m*τ[i]) + m))
```
3. I reorganized the statement slightly:
```julia
    I += (-next_exp_negτ * (m*(τ[i+1]-τ[i]) + S[i] + m) +  
          cur_exp_negτ * (S[i] + m))
```
4. I replaced `m*(τ[i+1]-τ[i])` with `(S[i+1] - S[i])`:
```julia
    I += (-next_exp_negτ * ((S[i+1] - S[i]) + S[i] + m) +  
          cur_exp_negτ * (S[i] + m))
```
5. After a final cancellation, we're left with:
```julia
    I += (-next_exp_negτ * (S[i+1] + m) +  
          cur_exp_negτ * (S[i] + m))
```

Using this result, you can rewrite the function I started with as:
```julia
function ray_transfer_integral(τ, S)
    I = 0
    next_exp_negτ = exp(-τ[1])
    for i in 1:length(τ)-1
        @inbounds m = (S[i+1] - S[i])/(τ[i+1] - τ[i])
        cur_exp_negτ = next_exp_negτ
        @inbounds next_exp_negτ = exp(-τ[i+1])
        @inbounds I += (-next_exp_negτ * (S[i+1] + m) +  
                        cur_exp_negτ * (S[i] + m))
    end
    I
end
```